app/estimate_square_size.go

```diff
@@ -1,204 +1,91 @@
 package app
 
 import (
-	"bytes"
 	"math"
-	"sort"
 
 	"github.com/celestiaorg/celestia-app/pkg/appconsts"
 	"github.com/celestiaorg/celestia-app/pkg/shares"
-	"github.com/cosmos/cosmos-sdk/client"
 	coretypes "github.com/tendermint/tendermint/types"
 )
 
-// prune removes txs until the set of txs will fit in the square of size
-// squareSize. It assumes that the currentShareCount is accurate. This function
-// is far from optimal because accurately knowing how many shares any given
-// set of transactions and its blob takes up in a data square that is following the
-// non-interactive default rules requires recalculating the entire square.
-// TODO: include the padding used by each blob when counting removed shares
-func prune(txConf client.TxConfig, txs []*parsedTx, currentShareCount, squareSize int) parsedTxs {
-	maxShares := squareSize * squareSize
-	if maxShares >= currentShareCount {
-		return txs
-	}
-	goal := currentShareCount - maxShares
-
-	removedContiguousShares := 0
-	contigBytesCursor := 0
-	removedBlobShares := 0
-	removedTxs := 0
-
-	// adjustContigCursor checks if enough contiguous bytes have been removed
-	// inorder to tally total contiguous shares removed
-	adjustContigCursor := func(l int) {
-		contigBytesCursor += l + shares.DelimLen(uint64(l))
-		if contigBytesCursor >= appconsts.ContinuationCompactShareContentSize {
-			removedContiguousShares += (contigBytesCursor / appconsts.ContinuationCompactShareContentSize)
-			contigBytesCursor = contigBytesCursor % appconsts.ContinuationCompactShareContentSize
-		}
-	}
-
-	for i := len(txs) - 1; (removedContiguousShares + removedBlobShares) < goal; i-- {
-		// this normally doesn't happen, but since we don't calculate the number
-		// of padded shares also being removed, its possible to reach this value
-		// should there be many small blobs, and we don't want to panic.
-		if i < 0 {
-			break
-		}
-		removedTxs++
-		if txs[i].msg == nil {
-			adjustContigCursor(len(txs[i].rawTx))
+// estimateSquareSize uses the provided block data to over estimate the square
+// size and the starting share index of non-reserved namespaces. The estimates
+// returned are liberal in the sense that we assume close to worst case and
+// round up.
+//
+// NOTE: The estimation process does not have to be perfect. We can overestimate
+// because the cost of padding TODO: cache and return the number of shares a
+// blob uses so we don't recalculate it later.
+func estimateSquareSize(txs []parsedTx) (squareSize uint64, nonreserveStart int) {
+	txSharesUsed := estimateCompactShares(appconsts.MaxSquareSize, txs)
+	blobSharesUsed := 0
+
+	for _, ptx := range txs {
+		if len(ptx.normalTx) != 0 {
```

Comment:
[for reviewers]

the estimation process has been significantly simplified by overestimating the number of compact shares used, and always using that estimate as the start of the sparse shares.